### PR TITLE
feat: add support for MariaDB SocketFactory connector

### LINKED ISSUE #1163: Support MariaDB JDBC Connector
Add support for https://mariadb.com/kb/en/about-mariadb-connector-j/

### PR DESCRIPTION
**Pull request by fzakaria**
## Change Description

Add support for MariaDB SocketFactory connector.

## Checklist

- [x] Make sure to open an issue as a 
  [bug/issue](https://github.com/GoogleCloudPlatform//issues/new/choose) 
  before writing your code!  That way we can discuss the change, evaluate 
  designs, and agree on the general idea.
- [x] Ensure the tests and linter pass
- [x] Appropriate documentation is updated (if necessary)

## Relevant issues:

- Fixes #1163

### CODE REVIEW

**Comment by fzakaria** [comment]
image
Relevant portion from the documentation.

**Comment by fzakaria** [comment]
Added tests that pass. I created a MySQL instance on CloudSQL.

Only annoying PITA is that a hostname and port **must be** specified for it to get to the point where it invokes the SocketFactory.

CC @tjbanghart

**Comment by fzakaria** [comment]
I tested IAM integration in addition to username & password.
Both work nicely.

File: jdbc/mariadb/src/main/java/com/google/cloud/sql/mariadb/SocketFactory.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+  @Override
+  public void setConfiguration(Configuration conf, String host) {
+    // Ignore the hostname
+    this.conf = conf;
+  }
+
+  public Socket createSocket() throws IOException {
+    try {
+      return CoreSocketFactory.connect(conf.nonMappedOptions());
```
Comment (fzakaria) [inline jdbc/mariadb/src/main/java/com/google/cloud/sql/mariadb/SocketFactory.java:53]: Relevant: https://github.com/mariadb-corporation/mariadb-connector-j/blob/master/src/main/java/org/mariadb/jdbc/client/impl/ConnectionHelper.java#L96

**Review by shubha-rajan — APPROVED** [review]

File: jdbc/mariadb/src/test/java/com/google/cloud/sql/mariadb/JdbcMariaDBIamAuthIntegrationTests.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+import org.junit.runners.JUnit4;
+
+
+@RunWith(JUnit4.class)
+public class JdbcMariaDBIamAuthIntegrationTests {
+  private static final String CONNECTION_NAME = System.getenv("MYSQL_CONNECTION_NAME");
+  private static final String DB_NAME = System.getenv("MYSQL_DB");
+  private static final String DB_USER = System.getenv("MYSQL_IAM_USER");
+  private static ImmutableList<String> requiredEnvVars = ImmutableList
+      .of("MYSQL_CONNECTION_NAME", "MYSQL_DB", "MYSQL_IAM_USER");
```
Comment (shubha-rajan) [inline jdbc/mariadb/src/test/java/com/google/cloud/sql/mariadb/JdbcMariaDBIamAuthIntegrationTests.java:51]:
```suggestion
      .of("MYSQL_IAM_CONNECTION_NAME", "MYSQL_DB", "MYSQL_IAM_USER");
```

File: jdbc/mariadb/src/test/java/com/google/cloud/sql/mariadb/JdbcMariaDBIamAuthIntegrationTests.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.Timeout;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+
+@RunWith(JUnit4.class)
+public class JdbcMariaDBIamAuthIntegrationTests {
+  private static final String CONNECTION_NAME = System.getenv("MYSQL_CONNECTION_NAME");
```
Comment (shubha-rajan) [inline jdbc/mariadb/src/test/java/com/google/cloud/sql/mariadb/JdbcMariaDBIamAuthIntegrationTests.java:47]:
```suggestion
  private static final String CONNECTION_NAME = System.getenv("MYSQL_IAM_CONNECTION_NAME");
```

**Pull request merged**